### PR TITLE
MSVC RTTI class name demangling

### PR DESCRIPTION
**Pull request by thestr4ng3r**
I am aware that there is already msvc demangling in bin and it would intuitively make sense to reuse that here. However while the class name format in rtti looks very similar to the one used for symbols normally, it is different.
Since msvc mangling is not officially documented and I wasn't able to find any other place where this kind of class name mangling was used by msvc other than rtti, I decided to add it to the rtti code.

Tests: https://github.com/radare/radare2-regressions/pull/1446

### CODE REVIEW

**Comment by XVilka** [comment]
Very good.

File: libr/anal/rtti.c [hunk 1]
```diff
@@ -2,6 +2,17 @@
 
 #include "r_anal.h"
 
+R_API char *r_anal_rtti_demangle_class_name(RAnal *anal, const char *name) {
+	RVTableContext context;
+	r_anal_vtable_begin (anal, &context);
+	if (context.abi == R_ANAL_CPP_ABI_MSVC) {
+		return r_anal_rtti_msvc_demangle_class_name (name);
+	} else {
```
Comment (radare) [inline libr/anal/rtti.c]: dont else after return

File: libr/anal/rtti_msvc.c [hunk 2]
```diff
@@ -357,6 +357,51 @@ static void rtti_msvc_print_base_class_descriptor_json(rtti_base_class_descripto
...
+ * => ClassA
+ * .?AVClassInInnerNamespace@InnerNamespace@OuterNamespace@@
+ * => OuterNamespace::InnerNamespace::AVClassInInnerNamespace
+ */
+R_API char *r_anal_rtti_msvc_demangle_class_name(const char *name) {
+	if (!name) {
+		return NULL;
+	}
+	size_t original_len = strlen (name);
+	if (original_len < 7
```
Comment (radare) [inline libr/anal/rtti_msvc.c]: is original_len > 2 other wise there can be an underflow in the strncmp below

File: libr/anal/rtti_msvc.c [hunk 3]
```diff
@@ -357,6 +357,51 @@ static void rtti_msvc_print_base_class_descriptor_json(rtti_base_class_descripto
...
+	const char *oc = name + original_len - 3;
+	size_t part_len = 0;
+	while (oc >= name + 4) {
+		if (*oc == '@') {
+			memcpy (c, oc + 1, part_len);
+			c += part_len;
+			*c++ = ':';
+			*c++ = ':';
+			part_len = 0;
+			oc--;
```
Comment (radare) [inline libr/anal/rtti_msvc.c]: move oc-- outside the if block. its done in both cases

File: libr/anal/rtti_msvc.c (hunk 2 above)
Comment (radare) [reply libr/anal/rtti_msvc.c]: oh nvm its ||. its ok

**Pull request merged**